### PR TITLE
Fix `pretraining_transforms` for `ImageEmbedder`

### PR DESCRIPTION
**Pull request by krshrimali**
## What does this PR do?

Attempts to fix https://github.com/PyTorchLightning/lightning-flash/issues/1185. Before this, `pretraining_transforms` was never called.

## Before submitting
- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the **[contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md)**, Pull Request section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes?
- [x] Did you write any **new necessary tests**? [not needed for typos/docs]
- [x] Did you verify **new and existing tests pass** locally with your changes?
- [x] If you made a notable change (that affects users), did you **update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)**?

## PR review
 - [x] Is this pull request **ready for review**? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by krshrimali** [comment]
Update: looks like it's still not called. WIP :(

File: tests/image/embedding/test_model.py [hunk 1]
```diff
@@ -51,7 +51,7 @@ def test_load_from_checkpoint_dependency_error():
 
 
 @pytest.mark.skipif(not (_TORCHVISION_AVAILABLE and _VISSL_AVAILABLE), reason="vissl not installed.")
-@pytest.mark.parametrize("backbone, training_strategy", [("resnet", "barlow_twins")])
+@pytest.mark.parametrize("backbone, training_strategy", [("vision_transformer", "dino")])
```
Comment (krshrimali) [inline tests/image/embedding/test_model.py:54]: This and other changes in the tests will be reverted, and we'll instead pass multiple args to test for multiple backbones and pretraining transforms. For now, relying on CI to give results as local GPU isn't working.

File: CHANGELOG.md [hunk 2]
```diff
@@ -20,6 +20,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
 
+- Fixed a bug where input transforms in the `ImageEmbedder` were not called. ([1196](https://github.com/PyTorchLightning/lightning-flash/pull/1196))
```
Comment (krshrimali) [inline CHANGELOG.md:23]:
```suggestion
- Fixed a bug where `pretraining_transforms` in the `ImageEmbedder` was never called. ([1196](https://github.com/PyTorchLightning/lightning-flash/pull/1196))
```

File: flash/image/embedding/model.py [hunk 3]
```diff
@@ -110,8 +113,17 @@ def __init__(
             learning_rate=learning_rate,
         )
 
+        @dataclass
+        class LambdaInputTransform(InputTransform):
```
Comment (ethanwharris) [inline flash/image/embedding/model.py:117]: Let's just import the one we have in `flash.core.data.io.input_transform` instead :smiley:

File: tests/image/embedding/test_model.py [hunk 4]
```diff
@@ -51,26 +51,35 @@ def test_load_from_checkpoint_dependency_error():
...
-@pytest.mark.parametrize("backbone, training_strategy", [("resnet", "barlow_twins")])
-def test_vissl_training(tmpdir, backbone, training_strategy):
+@pytest.mark.parametrize(
+    "backbone, training_strategy, head, pretraining_transform",
+    [
+        ("vision_transformer", "simclr", "simclr_head", "simclr_transform"),
+        ("vision_transformer", "dino", "dino_head", "dino_transform"),
+        ("vision_transformer", "barlow_twins", "simclr_head", "barlow_twins_transform"),
+        ("vision_transformer", "swav", "swav_head", "swav_transform"),
+    ],
```
Comment (ethanwharris) [inline tests/image/embedding/test_model.py:61]: Nice!

File: flash/core/adapter.py [hunk 5]
```diff
@@ -79,15 +78,6 @@ def __init__(self, adapter: Adapter, **kwargs):
...
 
-    @torch.jit.unused
-    @property
-    def input_transform(self) -> Optional[INPUT_TRANSFORM_TYPE]:
-        return self.adapter.input_transform
-
-    @input_transform.setter
-    def input_transform(self, input_transform: INPUT_TRANSFORM_TYPE) -> None:
-        self.adapter.input_transform = input_transform
-
```
Comment (ethanwharris) [inline flash/core/adapter.py:90]: Can we keep these?

**Review by ethanwharris — APPROVED** [review]
Awesome!!! LGTM, just a few small comments

File: flash/core/adapter.py (hunk 5 above)
Comment (krshrimali) [reply flash/core/adapter.py]: Sure! Done.

File: flash/image/embedding/model.py (hunk 3 above)
Comment (krshrimali) [reply flash/image/embedding/model.py]: Ah yes, done! 🎉

**Comment by krshrimali** [comment]
@ethanwharris - How to get rid of codecov failures? Can you please help with that?

Also, there are these tests failing: https://dev.azure.com/PytorchLightning/lightning%20Flash/_build/results?buildId=59765&view=logs&j=fb683405-d979-52da-6de9-2541dff429a6

```
# Failed while doing: nltk.download("punkt", quiet=True, force=False)
```

With an error:
```bash
Resource [93mpunkt[0m not found.
2022-03-03T04:45:07.0342026Z   Please use the NLTK Downloader to obtain the resource:
```
And the suggested fix from the CI:

```python
# suggestion from the error:
import nltk
nltk.download('punkt')
```

Should we try downloading `punkt` with the CI config for azure?

**Comment by krshrimali** [comment]
Thank you @ethanwharris for fixing the failures! All the tests pass now 🚀, regarding the DeepSource failures, I think we can safely ignore them for now?  (is there a way to auto-merge with these failures?)

I have a minor suggestion though: https://deepsource.io/gh/PyTorchLightning/lightning-flash/run/cc57fc8b-c459-4137-8c24-69f7ddc9b119/python/PYL-W0221 - I think deepsource is right here, and we can make `loss_fn, backbone, head, hooks` as kwargs (it won't be a BC breaking change) in a separate PR, to make sure that we override correctly. What do you think?

**Pull request merged**